<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [219]:
import pandas as pd
import psycopg2

In [220]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [221]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [222]:
query_3_1_count = '''select count(*) as vacancy_count
from vacancies
'''
df_vacancy_count = pd.read_sql_query(query_3_1_count, connection)
df_vacancy_count

C:\Users\user\AppData\Local\Temp\ipykernel_4784\1583263605.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacancy_count = pd.read_sql_query(query_3_1_count, connection)


vacancy_count
0          49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [223]:

query_3_2_count = '''select count(*) as employer_count
from employers
'''
df_employer_count = pd.read_sql_query(query_3_2_count, connection)
df_employer_count

C:\Users\user\AppData\Local\Temp\ipykernel_4784\152925328.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employer_count = pd.read_sql_query(query_3_2_count, connection)


employer_count
0           23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [224]:

query_3_3_count = '''select count(*) as area_count
from areas
'''
df_area_count = pd.read_sql_query(query_3_3_count, connection)
df_area_count

C:\Users\user\AppData\Local\Temp\ipykernel_4784\1335259410.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_area_count = pd.read_sql_query(query_3_3_count, connection)


area_count
0        1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [225]:
query_3_4_count = '''select count(*) as industry_count
from industries
'''
df_industry_count = pd.read_sql_query(query_3_4_count, connection)
df_industry_count

C:\Users\user\AppData\Local\Temp\ipykernel_4784\1101427843.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_industry_count = pd.read_sql_query(query_3_4_count, connection)


industry_count
0             294

***

 Выводы по предварительному анализу данных:
 1. В базе данных содержится 49197 вакансий (df_vacancy_count).
 2. Количество работодателей составляет 23501 (df_employer_count).
 3. В базе данных представлено 1362 региона (df_area_count).
 4. Количество сфер деятельности, указанных в базе, равно 294 (df_industry_count).

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [226]:
query_4_1 = """
SELECT 
    a.name AS area, 
    COUNT(v.id) AS cnt
FROM 
    areas a
LEFT JOIN 
    vacancies v ON a.id = v.area_id
GROUP BY 
    a.name
ORDER BY 
    cnt DESC
"""

In [227]:
df5 = pd.read_sql_query(query_4_1, connection)
df5

C:\Users\user\AppData\Local\Temp\ipykernel_4784\781545019.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5 = pd.read_sql_query(query_4_1, connection)


area   cnt
0                                Москва  5333
1                       Санкт-Петербург  2851
2                                 Минск  2112
3                           Новосибирск  2006
4                                Алматы  1892
...                                 ...   ...
1357                         Багаевская     0
1358                         Пролетарск     0
1359                       Седельниково     0
1360                             Инской     0
1361  Белоярский (Свердловская область)     0

[1362 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [228]:
query_4_2 = """
SELECT COUNT(*)
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
"""

In [229]:
df6 = pd.read_sql_query(query_4_2, connection)
df6

C:\Users\user\AppData\Local\Temp\ipykernel_4784\4153525677.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6 = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [230]:
query_4_3 = """
SELECT 
    ROUND(AVG(salary_from)) AS avg_salary_from,
    ROUND(AVG(salary_to)) AS avg_salary_to
FROM 
    vacancies
WHERE 
    salary_from IS NOT NULL OR salary_to IS NOT NULL
"""

In [231]:
df7 = pd.read_sql_query(query_4_3, connection)
df7

C:\Users\user\AppData\Local\Temp\ipykernel_4784\3399367520.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df7 = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [232]:
query_4_4 = """
SELECT 
    schedule,
    employment,
    COUNT(*) AS vacancy_count
FROM 
    vacancies
GROUP BY 
    schedule, employment
ORDER BY 
    vacancy_count DESC
"""

In [233]:
df8 = pd.read_sql_query(query_4_4, connection)
df8

C:\Users\user\AppData\Local\Temp\ipykernel_4784\4068426121.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df8 = pd.read_sql_query(query_4_4, connection)


schedule           employment  vacancy_count
0        Полный день     Полная занятость          35367
1   Удаленная работа     Полная занятость           7802
2      Гибкий график     Полная занятость           1593
3   Удаленная работа  Частичная занятость           1312
4     Сменный график     Полная занятость            940
5        Полный день           Стажировка            569
6     Вахтовый метод     Полная занятость            367
7        Полный день  Частичная занятость            347
8      Гибкий график  Частичная занятость            312
9        Полный день     Проектная работа            141
10  Удаленная работа     Проектная работа            133
11     Гибкий график           Стажировка            116
12    Сменный график  Частичная занятость            101
13  Удаленная работа           Стажировка             64
14     Гибкий график     Проектная работа             18
15    Сменный график           Стажировка             12
16    Вахтовый метод     Проектная работа              2
17    Сменный график     Проектная работа              1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [234]:
query_4_5 = """
SELECT 
    experience,
    COUNT(*) AS vacancy_count
FROM 
    vacancies
GROUP BY 
    experience
ORDER BY 
    vacancy_count ASC
"""

In [235]:
df9 = pd.read_sql_query(query_4_5, connection)
df9


C:\Users\user\AppData\Local\Temp\ipykernel_4784\3523035317.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df9 = pd.read_sql_query(query_4_5, connection)


experience  vacancy_count
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

# Выводы по детальному анализу вакансий

1. Наибольшее количество вакансий сосредоточено в крупных городах, таких как Москва и Санкт-Петербург, что подтверждается данными из `df5`.

2. Примерно половина вакансий имеет указанные зарплатные вилки (`df6`), а средние значения нижней и верхней границ зарплат составляют 71,065 и 110,537 соответственно (`df7`).

3. Большинство вакансий предполагает полный рабочий день и полную занятость, что видно из анализа данных `df8`.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [236]:
query_5_1 = """
    SELECT 
        e.name AS employer_name,
        COUNT(v.id) AS vacancy_count
    FROM 
        employers e
    LEFT JOIN 
        vacancies v ON e.id = v.employer_id
    GROUP BY 
        e.name
    ORDER BY 
        vacancy_count DESC
    LIMIT 5

"""

In [237]:
df9 = pd.read_sql_query(query_5_1, connection)
df9

C:\Users\user\AppData\Local\Temp\ipykernel_4784\3574445526.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df9 = pd.read_sql_query(query_5_1, connection)


employer_name  vacancy_count
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [238]:
query_5_2 ="""SELECT 
    a.name as area_name,
    COUNT( distinct e.id) as employers_count,
    COUNT(distinct v.id) as vacancies_count
FROM   
    areas a
LEFT JOIN employers e ON e.area = a.id
LEFT JOIN vacancies v ON v.area_id = a.id
where v.id is Null
GROUP BY 1
ORDER BY 2 desc, 2 DESC
"""

In [239]:
df10 = pd.read_sql_query(query_5_2, connection)
df10

C:\Users\user\AppData\Local\Temp\ipykernel_4784\187801456.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df10 = pd.read_sql_query(query_5_2, connection)


area_name  employers_count  vacancies_count
0                            Россия              410                0
1                         Казахстан              207                0
2                Московская область               75                0
3                Краснодарский край               19                0
4                          Беларусь               18                0
..                              ...              ...              ...
588                        Кувандык                0                0
589                        Куйбышев                0                0
590  Куйбышево (Ростовская область)                0                0
591                        Култаево                0                0
592                        Кульсары                0                0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [240]:
query_5_3 = """
SELECT 
    e.name AS employer_name,
    COUNT(DISTINCT v.area_id) AS region_count
FROM 
    employers e
JOIN 
    vacancies v ON e.id = v.employer_id
GROUP BY 
    e.name
ORDER BY 
    region_count DESC
"""

In [241]:
df11 = pd.read_sql_query(query_5_3, connection)
df11

C:\Users\user\AppData\Local\Temp\ipykernel_4784\3116903500.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df11 = pd.read_sql_query(query_5_3, connection)


employer_name  region_count
0                      Яндекс           181
1                  Ростелеком           152
2                  Спецремонт           116
3      Поляков Денис Иванович            88
4                   ООО ЕФИН             71
...                       ...           ...
14761                  UniSol             1
14762            UNISTORY LLC             1
14763                   UNIT6             1
14764     United Distribution             1
14765              UNITEL LLC             1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [242]:
query_5_4 = """
SELECT 
    COUNT(e.id) AS employer_count_without_industry
FROM 
    employers e
LEFT JOIN 
    employers_industries ei ON e.id = ei.employer_id
WHERE 
    ei.industry_id IS NULL
"""

In [243]:
df12 = pd.read_sql_query(query_5_4, connection)
df12

C:\Users\user\AppData\Local\Temp\ipykernel_4784\566815606.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df12 = pd.read_sql_query(query_5_4, connection)


employer_count_without_industry
0                             8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [244]:
query_5_5 = """
SELECT 
    e.name
FROM 
    employers e
JOIN 
    employers_industries ei ON e.id = ei.employer_id
GROUP BY 
    e.name
HAVING 
    COUNT(ei.industry_id) = 4
ORDER BY 
    e.name
OFFSET 2 LIMIT 1
"""

In [245]:
df13 = pd.read_sql_query(query_5_5, connection)
df13

C:\Users\user\AppData\Local\Temp\ipykernel_4784\1626685551.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df13 = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [246]:
query_5_6 = """
SELECT 
    COUNT(DISTINCT e.id) AS employer_count
FROM 
    employers e
JOIN 
    employers_industries ei ON e.id = ei.employer_id
JOIN 
    industries i ON ei.industry_id = i.id
WHERE 
    i.name = 'Разработка программного обеспечения'
"""

In [247]:
df14 = pd.read_sql_query(query_5_6, connection)
df14

C:\Users\user\AppData\Local\Temp\ipykernel_4784\434231212.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df14 = pd.read_sql_query(query_5_6, connection)


employer_count
0            3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [248]:
from urllib.parse import quote


# Кодируем URL для обработки не-ASCII символов
url = "https://ru.wikipedia.org/wiki/Города-миллионеры_России"
encoded_url = quote(url, safe=':/')

# Читаем HTML-таблицу
tables = pd.read_html(encoded_url, encoding='utf-8')

# Извлекаем первую таблицу
million_cities_df = tables[0]

# Получаем список городов
million_cities = million_cities_df['Город'].tolist()
million_cities = [city.strip() for city in million_cities] # Убираем лишние пробелы
million_cities = tuple(million_cities)
print(million_cities)


('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [249]:
query_5_7 = f'''WITH temp AS(
                     select a.name,
                     COUNT (*) cnt
                FROM 
                     vacancies v
                LEFT JOIN 
                     areas a ON a.id = v.area_id
                LEFT JOIN 
                     employers e ON e.id = v.employer_id
                WHERE e.name = 'Яндекс'
                      AND a.name in {million_cities}
                GROUP BY 1
                ORDER BY 2
               )
                SELECT *  FROM temp
                UNION ALL
                SELECT 
                     'Total', 
                     sum(cnt)  
                FROM temp'''

In [250]:
df15 = pd.read_sql_query(query_5_7, connection)
df15

C:\Users\user\AppData\Local\Temp\ipykernel_4784\361924777.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df15 = pd.read_sql_query(query_5_7, connection)


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

# Выводы по анализу работодателей

1. Наибольшее количество вакансий размещено такими работодателями, как "Яндекс", "Ростелеком", "Тинькофф", "СБЕР" и "Газпром нефть" (данные из `df9`).

2. Среди регионов, в которых нет вакансий, наибольшее количество работодателей находится в России и Казахстане (данные из `df10`).

3. Работодатели, такие как "Яндекс" и "Ростелеком", публикуют вакансии в большом количестве регионов, что свидетельствует о широкой географической представленности (данные из `df11`).

4. У 8419 работодателей не указана сфера деятельности, что может затруднять анализ их профиля (данные из `df12`).

5. Компания "2ГИС" занимает третье место в алфавитном списке компаний, у которых указано четыре сферы деятельности (данные из `df13`).

6. 3553 работодателя указали в качестве сферы деятельности "Разработка программного обеспечения", что подчеркивает популярность этой отрасли (данные из `df14`).

7. Вакансии компании "Яндекс" представлены во всех городах-миллионниках, что подтверждает её активное присутствие на рынке труда (данные из `df15`).

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [251]:
query_6_1 = """
SELECT COUNT(*) AS data_related_vacancies
FROM vacancies
WHERE LOWER(name) LIKE '%data%' OR LOWER(name) LIKE '%данн%'
"""

In [252]:
df16 = pd.read_sql_query(query_6_1, connection)
df16

C:\Users\user\AppData\Local\Temp\ipykernel_4784\3068151413.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df16 = pd.read_sql_query(query_6_1, connection)


data_related_vacancies
0                    1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [253]:
query_6_2 = """
SELECT COUNT(*) AS junior_ds_vacancies
FROM vacancies
WHERE 
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
    )
    AND (
        LOWER(name) LIKE '%junior%' OR
        experience = 'Нет опыта' OR
        employment = 'Стажировка'
    )
"""

In [254]:
df17 = pd.read_sql_query(query_6_2, connection)
df17

C:\Users\user\AppData\Local\Temp\ipykernel_4784\2941832824.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df17 = pd.read_sql_query(query_6_2, connection)


junior_ds_vacancies
0                   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [255]:
query_6_3 = """
SELECT COUNT(*) AS ds_vacancies_with_sql
FROM vacancies
WHERE 
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
    )
    AND (
        LOWER(key_skills) LIKE '%sql%' OR
        LOWER(key_skills) LIKE '%postgres%'
    )
"""

In [256]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_4784\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


ds_vacancies_with_sql
0                    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [257]:
query_6_4 = """
SELECT COUNT(*) AS ds_vacancies_with_python
FROM vacancies
WHERE 
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT iLIKE '%html%') OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%'
    )
    AND LOWER(key_skills) LIKE '%python%'
"""

In [258]:
df18 = pd.read_sql_query(query_6_4, connection)
df18

C:\Users\user\AppData\Local\Temp\ipykernel_4784\3636623983.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df18 = pd.read_sql_query(query_6_4, connection)


ds_vacancies_with_python
0                       357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [259]:
query_6_5 = """
SELECT ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2) AS avg_key_skills
FROM vacancies
WHERE 
    (
        LOWER(name) iLIKE '%data scientist%' OR
        LOWER(name) iLIKE '%data science%' OR
        LOWER(name) iLIKE '%исследователь данных%' OR
        (LOWER(name) LIKE '%ML%' AND LOWER(name) NOT iLIKE '%html%') OR
        LOWER(name) iLIKE '%machine learning%' OR
        LOWER(name) iLIKE '%машинн%обучен%'
    )
    AND key_skills IS NOT NULL
"""

In [260]:
df19 = pd.read_sql_query(query_6_5, connection)
df19

C:\Users\user\AppData\Local\Temp\ipykernel_4784\4117865023.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df19 = pd.read_sql_query(query_6_5, connection)


avg_key_skills
0            6.44

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [261]:
query_6_6 = """
SELECT
  v.experience, ROUND(AVG(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0) AS avg_salary
FROM
  vacancies v
WHERE
(
       (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)

GROUP BY v.experience
ORDER BY 1
"""

In [262]:
# текст запроса
df20 = pd.read_sql_query(query_6_6, connection)
df20



C:\Users\user\AppData\Local\Temp\ipykernel_4784\2364795686.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df20 = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0         Более 6 лет         NaN
1           Нет опыта     74643.0
2  От 1 года до 3 лет    145236.0
3       От 3 до 6 лет    243115.0

***

# Выводы по предметному анализу

1. В базе данных содержится 1771 вакансия, связанная с данными, что свидетельствует о значительном спросе на специалистов в этой области (`df16`).

2. Вакансий для начинающих дата-сайентистов (Junior) всего 51, что указывает на высокие требования к квалификации даже для начального уровня (`df17`).

3. SQL и Postgres являются важными навыками для дата-сайентистов, так как 229 вакансий требуют их знания (`df`).

4. Python является наиболее популярным языком программирования среди вакансий для дата-сайентистов, упоминается в 357 вакансиях (`df18`).

5. Среднее количество ключевых навыков, указанных в вакансиях для дата-сайентистов, составляет 6.44, что подчеркивает высокие требования к квалификации специалистов (`df19`).

6. Средняя зарплата для дата-сайентистов с опытом от 3 до 6 лет составляет 243,115, что делает этот сегмент наиболее привлекательным для специалистов (`df20`).

# Общий вывод по проекту

In [263]:
print("Основные выводы по результатам анализа вакансий HeadHunter:")

print("\n1. Общая ситуация на рынке труда:")
print(f"- Всего в базе данных {df_vacancy_count.iloc[0,0]} вакансий и {df_employer_count.iloc[0,0]} работодателей.")
print(f"- Наибольшее количество вакансий сосредоточено в Москве ({df5.iloc[0,1]} вакансий) и Санкт-Петербурге ({df5.iloc[1,1]} вакансий).")

print("\n2. Анализ зарплатных предложений:")
print(f"- Зарплатные вилки указаны примерно в {round(df6.iloc[0,0] / df_vacancy_count.iloc[0,0] * 100)}% вакансий.")
print(f"- Средняя предлагаемая зарплата 'от': {df7.iloc[0,0]}, средняя предлагаемая зарплата 'до': {df7.iloc[0,1]}.")

print("\n3. Тип занятости и опыт:")
print(f"- Наиболее востребованный тип занятости: Полный день и Полная занятость ({df8.iloc[0,2]} вакансий).")
print(f"- Больше всего вакансий с опытом работы от 1 года до 3 лет.")

print("\n4. Работодатели:")
print(f"- Топ 5 работодателей по количеству вакансий: {', '.join(df9['employer_name'].tolist())}.")
print(f"- {df12.iloc[0,0]} работодателей не указали сферу деятельности.")

print("\n5. Data Science:")
print(f"- Найдено {df16.iloc[0,0]} вакансий, связанных с Data Science.")
print(f"- Python является наиболее популярным навыком, требуемым в {df18.iloc[0,0]} вакансиях DS.")
print(f"- Среднее количество ключевых навыков в вакансиях DS: {df19.iloc[0,0]}.")
print(f"- Средняя зарплата для DS с опытом работы от 3 до 6 лет: {df20[df20['experience'] == 'От 3 до 6 лет']['avg_salary'].iloc[0]}.")

print("\nДополнительные направления для исследования:")
print("- Анализ требований к образованию в вакансиях DS.")
print("- Исследование динамики изменения количества вакансий и зарплат в зависимости от месяца/года.")
print("- Сравнение требований и предложений для DS в разных регионах.")

connection.close()


Основные выводы по результатам анализа вакансий HeadHunter:

1. Общая ситуация на рынке труда:
- Всего в базе данных 49197 вакансий и 23501 работодателей.
- Наибольшее количество вакансий сосредоточено в Москве (5333 вакансий) и Санкт-Петербурге (2851 вакансий).

2. Анализ зарплатных предложений:
- Зарплатные вилки указаны примерно в 49% вакансий.
- Средняя предлагаемая зарплата 'от': 71065.0, средняя предлагаемая зарплата 'до': 110537.0.

3. Тип занятости и опыт:
- Наиболее востребованный тип занятости: Полный день и Полная занятость (35367 вакансий).
- Больше всего вакансий с опытом работы от 1 года до 3 лет.

4. Работодатели:
- Топ 5 работодателей по количеству вакансий: Яндекс, Ростелеком, Тинькофф, СБЕР, Газпром нефть.
- 8419 работодателей не указали сферу деятельности.

5. Data Science:
- Найдено 1771 вакансий, связанных с Data Science.
- Python является наиболее популярным навыком, требуемым в 357 вакансиях DS.
- Среднее количество ключевых навыков в вакансиях DS: 6.44.
- Средня

В ходе анализа рынка труда HeadHunter было выявлено, что наибольшее количество вакансий сосредоточено в крупных городах, таких как Москва и Санкт-Петербург, при этом наиболее востребованными являются специалисты с опытом работы от 1 до 3 лет на полный рабочий день.
Зарплатные предложения указаны примерно в половине вакансий, а наиболее популярным навыком для Data Science является Python, что подчеркивает его важность в данной области.
Средняя зарплата для DS с опытом работы от 3 до 6 лет является наиболее привлекательной, однако стоит отметить, что значительное число работодателей не указывают сферу деятельности, что может затруднять анализ их профиля.